airbyte-integrations/connectors/source-bigquery/source_bigquery/streams.py

```diff
@@ -500,23 +871,53 @@ def _updated_state(self, current_stream_state: MutableMapping[str, Any], latest_
         latest_record_state = latest_record[self.cursor_field]
         stream_state = current_stream_state.get(self.cursor_field)
         if stream_state:
-            self._cursor = max(latest_record_state, stream_state)
+            max_date = max(datetime.strptime(latest_record_state, '%Y-%m-%dT%H:%M:%S.%f%z'), datetime.strptime(stream_state, '%Y-%m-%dT%H:%M:%S.%f%z'))
+            self._cursor = max_date.isoformat(timespec='microseconds')
         else:
             self._cursor = latest_record_state
-        self.state = {self.cursor_field: self._cursor} 
-        self.logger.info(f"current state of {self.name} is {self.state}")
-
+        self.state = {self.cursor_field: self._cursor}
         if datetime.now() >= self._checkpoint_time + timedelta(minutes=15):
             self.checkpoint(self.name, self.state, self.namespace)
             self._checkpoint_time = datetime.now()
         return self.state
 
+    def _extract_borders(self):
+        if not self.start_time and not self.end_time:
+            cursor_column = next(key for key, value in CHANGE_FIELDS.items() if value == self.cursor_field)
+            first_record = CHTableQueryRecord(self.project_id, self.stream_name, "ASC", cursor_column, self.where_clause, authenticator=self._auth)
+            last_record = CHTableQueryRecord(self.project_id, self.stream_name, "DESC", cursor_column, self.where_clause, authenticator=self._auth)
+            self.start_time = next(first_record.read_records(sync_mode=SyncMode.full_refresh)).get(self.cursor_field, None)
+            self.end_time = next(last_record.read_records(sync_mode=SyncMode.full_refresh)).get(self.cursor_field, None)
+            if isinstance(self.start_time, str):
+                self.start_time = datetime.strptime(self.start_time, '%Y-%m-%dT%H:%M:%S.%f%z')
+            if isinstance(self.end_time, str):
+                self.end_time = datetime.strptime(self.end_time, '%Y-%m-%dT%H:%M:%S.%f%z')
+        return self.start_time, self.end_time
+    
+    def _chunk_dates(self, start_date: datetime, end_date: datetime, table_start: datetime) -> Iterable[Tuple[datetime, datetime]]:
+        step = timedelta(minutes=self.slice_range)
+        new_start_date = start_date
+        if start_date == end_date:
+            yield start_date, start_date + step
+            return
+        while new_start_date < end_date+timedelta(seconds=1):
+            before_date = min(end_date+timedelta(seconds=1), new_start_date + step)
+            if table_start and before_date < table_start:
+                before_date = table_start + step
+            yield new_start_date, before_date
+            new_start_date = before_date
+
     def stream_slices(self, stream_state: Mapping[str, Any] = None, cursor_field=None, sync_mode=None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        start_time, end_time = self._extract_borders()
+        default_start = self.fallback_start #max(start_time, self.fallback_start)
```

Comment:
Why is there a comment here?

Reply:
It shouldn't be, I will remove it right away.

Reply:
@maxi297 
fixed here https://github.com/artefactory/airbyte/pull/30

Reply:
@maxi297 
fixed here https://github.com/artefactory/airbyte/pull/30